cscs-checks/system/slurm/src/eatmemory_mpi.c

```diff
@@ -0,0 +1,173 @@
+      cscs_read_proc_meminfo(i);
+    }
+    short *buffer = malloc(sizeof(char) * chunk);
+    if (buffer == NULL) {
+      return false;
+    }
+    memset(buffer, 0, chunk);
+  }
+  return true;
+}
+
+int cscs_read_proc_meminfo(int i) {
+  FILE *fp;
+  meminfo_t meminfo[] = {{"MemTotal:", 0},     {"MemFree:", 0},
+                         {"Cached:", 0},       {"SwapCached:", 0},
+                         {"SwapTotal:", 0},    {"SwapFree:", 0},
+                         {"MemAvailable:", 0}, {NULL, 0}};
+  fp = fopen(PROC_FILE, "r");
+  if (!fp) {
+    printf("Cannot read %s", PROC_FILE);
+    return -1;
+  }
+  char buf[80];
+  while (fgets(buf, sizeof(buf), fp)) {
+    int i;
+    for (i = 0; meminfo[i].str; i++) {
+      size_t len = strlen(meminfo[i].str);
+      if (!strncmp(buf, meminfo[i].str, len)) {
+        char *ptr = buf + len + 1;
+        while (isspace(*ptr))
+          ptr++;
+        sscanf(ptr, "%u kB", &meminfo[i].val);
+      }
+    }
+  }
+  fclose(fp);
+
+  printf("Mem: total: %u GB, free: %u GB, avail: %u GB, using: %u GB\n",
+         meminfo[MEMTOTAL].val / 1048576, meminfo[MEMFREE].val / 1048576,
+         meminfo[MEMAVAIL].val / 1048576,
+         (meminfo[MEMTOTAL].val - meminfo[MEMAVAIL].val) / 1048576);
+  return 0;
+}
+
+int main(int argc, char *argv[]) {
+  int rank, mpi_size;
+  MPI_Init(&argc, &argv);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+#ifdef MEMORY_PERCENTAGE
+  if (rank == 0) {
+    printf("Currently total memory: %zd\n", getTotalSystemMemory());
+    printf("Currently avail memory: %zd\n", getFreeSystemMemory());
+  }
+#endif
+  int i;
+  for (i = 0; i < argc; i++) {
+    char *arg = argv[i];
+    if (strcmp(arg, "-h") == 0 || strcmp(arg, "-?") == 0 || argc == 1) {
+      printf("Usage: eatmemory <size>\n");
+      printf("Size can be specified in megabytes or gigabytes in the following "
+             "way:\n");
+      printf("#          # Bytes      example: 1024\n");
+      printf("#M         # Megabytes  example: 15M\n");
+      printf("#G         # Gigabytes  example: 2G\n");
+#ifdef MEMORY_PERCENTAGE
+      printf("#%%         # Percent    example: 50%%\n");
+#endif
+      printf("\n");
+    } else if (i > 0) {
+      int len = strlen(arg);
+      char unit = arg[len - 1];
+      long size = -1;
+      int chunk = 268435456; // = 256M
+      // int chunk=536870912; // = 512M
+      // int chunk=1073741824; // = 1G
+      if (!isdigit(unit)) {
+        if (unit == 'M' || unit == 'G') {
+          arg[len - 1] = 0;
+          size = atol(arg) * (unit == 'M' ? 1024 * 1024 : 1024 * 1024 * 1024);
+        }
+#ifdef MEMORY_PERCENTAGE
+        else if (unit == '%') {
+          size = (atol(arg) * (long)getFreeSystemMemory()) / 100;
+        }
+#endif
+        else {
+          printf("Invalid size format\n");
+          exit(0);
+        }
+      } else {
+        size = atoi(arg);
+      }
+
+      if (rank == 0) {
+        cscs_read_proc_meminfo(i);
```

Comment:
If you have edited this file, you should mention it at the beginning, that you are also an author.

Reply:
It came from the problem we had with: https://github.com/eth-cscs/reframe/pull/1613#issuecomment-731994574
Do you have an example of what you want ?

Reply:
Why did you have to change the code? I think the problem in that comment was due to how you were computing the sanity function. As I proposed, you go without a hook.

Reply:
refers to https://github.com/eth-cscs/reframe/pull/1613#discussion_r533725483